cpp/src/io/parquet/page_enc.cu

```diff
@@ -190,37 +194,28 @@ __global__ void __launch_bounds__(block_size) gpuInitPageFragments(PageFragment
   size_type nvals           = s->frag.num_leaf_values;
   size_type start_value_idx = s->start_value_idx;
 
-  size_type validity_offset = (s->col.nesting_offsets == nullptr) ? s->col.column_offset : 0;
   for (uint32_t i = 0; i < nvals; i += block_size) {
-    const uint32_t *valid = s->col.valid_map_base;
-    uint32_t val_idx      = start_value_idx + i + t;
-    uint32_t is_valid =
-      (i + t < nvals && val_idx < s->col.num_values)
-        ? (valid)
-            ? (valid[(val_idx + validity_offset) / 32] >> ((val_idx + validity_offset) % 32)) & 1
-            : 1
-        : 0;
+    uint32_t val_idx  = start_value_idx + i + t + s->col.leaf_column_offset;
```

Comment:
I feel like there's a problem with storing `leaf_column_offset` separately and applying it after getting the `start_value_idx`. This would mean that fragments are not guaranteed to start and stop at row boundaries, which would be required for DataPageHeaderV2 which requires pages start and stop at row boundaries.

I'd suggest that `leaf_column_offset` be implicitly applied during the conversion of row index to value index. This won't break dictionary encoding because you can still use column_offset for that and dictionary encoding is not supported on list columns for now anyway. For all other encoding purposes, this approach would work.

You can probably also store `start_value_idx` as a member of page and use it along with `page.num_leaf_values` to get the exact bounds of indices in the leaf column that belong in that page.